modules/tests/src/test/scala/dev/profunktor/redis4cats/TestScenarios.scala

```diff
@@ -575,4 +578,49 @@ trait TestScenarios { self: FunSuite =>
       _ <- IO(assert(c4 > 0, "merged hyperloglog should think it has more than 0 items in"))
     } yield ()
   }
+
+  def keyPatternSubScenario(client: RedisClient): IO[Unit] = {
+    import dev.profunktor.redis4cats.effect.Log.NoOp._
+    import scala.language.postfixOps
+
+    val pattern = "__keyevent*__:*"
+    val key     = "somekey"
+    val resources = for {
+      ref <- Resource.eval(Ref.of[IO, Option[RedisPatternEvent[String, String]]](None))
+      commands <- Redis[IO].fromClient(client, RedisCodec.Utf8)
+      sub <- PubSub.mkSubscriberConnection[IO, String, String](client, RedisCodec.Utf8)
+      stream <- Resource.pure(sub.psubscribe(RedisPattern(pattern)))
+      _ <- stream.foreach(output => ref.set(Some(output))).compile.drain.background
+      _ <- Resource.eval(commands.set(key, ""))
+      _ <- Resource.eval(commands.expire(key, 1 seconds))
+    } yield ref.get
+    resources.use(get => get.iterateUntil(_.isDefined).timeout(2 seconds)).map { result =>
```

Comment:
Timeouts in tests usually end up becoming flaky. If there's a better way to test this, that would be preferable.

Reply:
I've been trying to come up with a better way of testing it, but I can't come up with one. I have worked around it with a sleep for now.

We need to start pulling prior to publishing the message, otherwise we won't get the published message. So I start the pulling of the stream on another fiber and I could wait until resource acquisition/stream initialization on the fiber is done prior to proceeding. But that does not guarantee that we are actually subscribed. In principle, the subscription is not active until redis responds with `subscribe`. Today, we do not wait until redis responds, so there is an interval between the stream having started and the subscription actually being active. This is what the `sleep` solves in the tests.

We could change the stream implementation so it blocks the resource acquisition in `PubSubInternals.channel` and `PubSubInternals.pattern` until the `subscribe` event is received. I would be happy to implement this (I think this behavior makes more sense). But it is a change in behavior from how redis4cats work today.

What are your thoughts on this?

Reply:
Sorry for the long delay :-( 

We could add a `subscribed` method as suggested to `PubSubCommands`, but that does require quite a bit of care to handle correctly... And I'm starting to think that maybe we should just not handle it.

If we do add a `subscribed` method to `PubSubCommands` we need to consider what happens if we start two subscriptions to the same channel. Should `PubSubCommands` defer to the first or the second one? What if one of them is cancelled?
Having `subscribed` seems to imply also having a `psubscribed`, but patterns have a hierarchy of sorts, unlike channels. Does `psubscribed(Pattern("f*"))` imply `psubscribed(Pattern("foo*"))`?

Also, for newcomers, it might not be obvious that something like 
```
for {
  stream <- pubsub.subscribe(RedisChannel("foo"))
  _ <- pubsub.subscribed(RedisChannel("foo"))
  // do something with stream
} yield ()
```
will block forever, since the `Stream` won't emit the `subscribe` command until it is pulled, hence `.subscribed(RedisChannel("foo"))` will never complete.

For the above, I'm starting to favor simply adding some retries to the tests, which I have done. Please let me know what you think :-)

Reply:
Actually, even with the retries, the test hangs sometimes, especially on CI.
It seems that there is some race condition in the listener. I can't explain why, but what I observe is:
1. Test runs fine and gets to releasing resource
2. Test tries to release resources (including the subscription)
3. sub listener is removed (see `PubSubInternals#pattern`)
4. dispatcher is closed/stopped (see `PubSubInternals#pattern`)
5. Listener from 3 receives one more event and tries to dispatch it
6. Dispatch call blocks forever (<-- I assume this prevent the connection cleanup as well)
7. Test times out

I think something weird is happening with the dispatcher. I have tried dispatching on a closed dispatcher and that just throws an error. But we are not getting an error in these tests. Instead, it just block indefinitely.

Using either `unsafeRunAndForget` or `unsafeRunTimed` in the listener makes the test pass, but I'm not sure if we should be using either of those. Especially since I don't fully understand what the issue is.
I'm hoping that you do :crossed_fingers: